From 4156e7b0de6c5ea6598b967cb61aecb0b42a1cb1 Mon Sep 17 00:00:00 2001
From: Volodymyr Pavlenko <pavlenko.vv@pm.me>
Date: Wed, 4 Mar 2026 12:35:14 +0530
Subject: [PATCH 1/3] tlvf: add eVapType and carry VAP type in WSC and joined
 notification

Changes:
- Add TLVF enum class eVapType (AutoGenerated) and YAML definition.
- Extend WSC EncryptedSettingsPayload with a vap_type field (default OTHER).
- Carry vap_type in WSC M2 via an Airties vendor extension subelement.
- Add sVapType/sVapTypesList and include vap_type_list in ACTION_APMANAGER_JOINED_NOTIFICATION.
- Store vap_type in bcl wireless settings (sBssInfoConf) and in bwl VAPElement.
- Add helpers to map vap type <-> string and custom_alias <-> vap_type, and read CustomAlias via WBAPI.

Related PPM-3464, PPM-3542

Signed-off-by: Volodymyr Pavlenko <pavlenko.vv@pm.me>
Signed-off-by: Stephan Raj Ignatious Durairaj <signatio@qti.qualcomm.com>
---
 .../bcl/include/bcl/son/son_wireless_utils.h  | 11 +++
 .../bcl/source/son/son_wireless_utils.cpp     | 26 +++++++
 .../bwl/include/bwl/base_wlan_hal_types.h     |  3 +
 common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp   |  8 +-
 common/beerocks/bwl/whm/base_wlan_hal_whm.cpp |  8 ++
 .../tlvf/beerocks_message_apmanager.h         |  4 +
 .../beerocks/tlvf/beerocks_message_common.h   | 25 +++++++
 .../tlvf/beerocks_message_apmanager.cpp       | 26 +++++++
 .../tlvf/beerocks_message_apmanager.yaml      |  2 +
 .../tlvf/beerocks_message_common.yaml         | 13 +++-
 .../platform/bpl/cfg/dm/bpl_cfg_wifi.cpp      |  8 ++
 .../platform/wbapi/include/wbapi_utils.h      | 25 +++++++
 framework/platform/wbapi/wbapi_utils.cpp      | 40 ++++++++++
 .../include/tlvf/WSC/WSC_Attributes.h         |  5 +-
 .../include/tlvf/common/eVapType.h            | 73 ++++++++++++++++++
 .../src/tlvf/WSC/WSC_Attributes.cpp           | 13 ++++
 framework/tlvf/src/include/tlvf/WSC/m2.h      | 37 +++++++++-
 framework/tlvf/src/src/WSC/m2.cpp             | 74 +++++++++++++------
 .../tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml    |  6 +-
 framework/tlvf/yaml/tlvf/common/eVapType.yaml | 15 ++++
 20 files changed, 389 insertions(+), 33 deletions(-)
 create mode 100644 framework/tlvf/AutoGenerated/include/tlvf/common/eVapType.h
 create mode 100644 framework/tlvf/yaml/tlvf/common/eVapType.yaml

diff --git a/common/beerocks/bcl/include/bcl/son/son_wireless_utils.h b/common/beerocks/bcl/include/bcl/son/son_wireless_utils.h
index c2763c06a..39a315b2c 100644
--- a/common/beerocks/bcl/include/bcl/son/son_wireless_utils.h
+++ b/common/beerocks/bcl/include/bcl/son/son_wireless_utils.h
@@ -16,6 +16,7 @@
 #include <tlvf/WSC/eWscAuth.h>
 #include <tlvf/WSC/eWscEncr.h>
 #include <tlvf/WSC/eWscVendorExt.h>
+#include <tlvf/common/eVapType.h>
 
 #include <array>
 #include <deque>
@@ -302,6 +303,7 @@ public:
         bool profile1_backhaul_sta_association_disallowed = false;
         bool profile2_backhaul_sta_association_disallowed = false;
         bool hidden_ssid                                  = false;
+        eVapType vap_type                                 = eVapType::OTHER;
         std::string mld_id                                = "";
         bool bSTA                                         = false;
         std::string backhaul_wps_ssid;
@@ -480,6 +482,15 @@ public:
     static std::string wsc_to_bwl_encryption(WSC::eWscEncr enctype);
     static beerocks::eBssType wsc_to_bwl_bss_type(WSC::eWscVendorExtSubelementBssType bss_type);
     static std::list<uint8_t> string_to_wsc_oper_class(const std::string &operating_class);
+
+    /*
+     * @brief Converts string to vap_type
+     *
+     * @param type - the string (e.g., "home", "guest").
+     * @return string - eVapType enum
+     **/
+    static eVapType string_to_vap_type(const std::string &type);
+
     /**
      * @brief Get the vht central frequency object
      * 
diff --git a/common/beerocks/bcl/source/son/son_wireless_utils.cpp b/common/beerocks/bcl/source/son/son_wireless_utils.cpp
index 36db5714e..cd61c3028 100644
--- a/common/beerocks/bcl/source/son/son_wireless_utils.cpp
+++ b/common/beerocks/bcl/source/son/son_wireless_utils.cpp
@@ -1789,6 +1789,32 @@ std::list<uint8_t> wireless_utils::string_to_wsc_oper_class(const std::string &o
     return {};
 }
 
+eVapType wireless_utils::string_to_vap_type(const std::string &type)
+{
+    if (type.empty()) {
+        LOG(WARNING) << "string_to_vap_type: type is empty, returning eVapType::OTHER";
+        return eVapType::OTHER;
+    }
+
+    if (type == "home") {
+        return eVapType::HOME;
+    } else if (type == "guest") {
+        return eVapType::GUEST;
+    } else if (type == "video") {
+        return eVapType::VIDEO;
+    } else if (type == "backhaul") {
+        return eVapType::BACKHAUL;
+    } else if (type == "hotspot") {
+        return eVapType::HOTSPOT;
+    } else if (type == "staff") {
+        return eVapType::STAFF;
+    } else if (type == "isolated") {
+        return eVapType::ISOLATED;
+    } else /* if type == "other" */ {
+        return eVapType::OTHER;
+    }
+}
+
 bool wireless_utils::is_channel_in_operating_class(uint8_t operating_class, uint8_t channel)
 {
     auto channel_set = operating_class_to_channel_set(operating_class);
diff --git a/common/beerocks/bwl/include/bwl/base_wlan_hal_types.h b/common/beerocks/bwl/include/bwl/base_wlan_hal_types.h
index f67bc09e8..75aa03681 100644
--- a/common/beerocks/bwl/include/bwl/base_wlan_hal_types.h
+++ b/common/beerocks/bwl/include/bwl/base_wlan_hal_types.h
@@ -13,6 +13,8 @@
 #include <bcl/beerocks_message_structs.h>
 #include <bcl/network/net_struct.h>
 
+#include <tlvf/common/eVapType.h>
+
 #include <stdint.h>
 
 #include <array>
@@ -67,6 +69,7 @@ struct VAPElement {
     int8_t mld_id                                     = -1;
     int8_t num_links                                  = -1;
     std::string ap_mld_mac;
+    eVapType vap_type = eVapType::OTHER;
 
     bool operator==(const VAPElement &other) const
     {
diff --git a/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp b/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp
index 1da363b59..8ef95bc00 100644
--- a/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp
+++ b/common/beerocks/bwl/whm/ap_wlan_hal_whm.cpp
@@ -485,15 +485,15 @@ bool ap_wlan_hal_whm::update_vap_credentials(
 {
     LOG(DEBUG) << "updating vap credentials of radio " << get_iface_name()
                << " and bridge=" << bridge_ifname;
-    bool ret;
+    bool ret          = false;
     int new_vap_index = m_radio_info.available_vaps.size();
 
-    for (auto bss_info_conf : bss_info_conf_list) {
+    for (const auto &bss_info_conf : bss_info_conf_list) {
         std::string wifi_vap_path, wifi_ssid_path;
         std::string ifname = "new_interface";
 
-        auto bssid = tlvf::mac_to_string(bss_info_conf.bssid);
-        int vap_id = get_vap_id_with_mac(bssid);
+        const auto bssid = tlvf::mac_to_string(bss_info_conf.bssid);
+        int vap_id       = get_vap_id_with_mac(bssid);
 
         if (!check_vap_id(vap_id) || (bssid == beerocks::net::network_utils::WILD_MAC_STRING)) {
             LOG(DEBUG) << "create new vap for wildcard bssid";
diff --git a/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp b/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
index d89550326..70549ace9 100644
--- a/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
+++ b/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
@@ -1476,8 +1476,16 @@ bool base_wlan_hal_whm::refresh_vap_info(int id, const AmbiorixVariant &ap_obj)
                     vap_element.backhaul = true;
                 }
             }
+
             m_ambiorix_cl.resolve_path(wbapi_utils::search_path_ap_by_iface(ifname),
                                        vap_extInfo.path);
+
+            // Reading CustomAlias and deducing VAP type
+            const std::string custom_alias = wbapi_utils::get_custom_alias(ap_obj);
+            vap_element.vap_type           = wbapi_utils::vap_type_from_custom_alias(custom_alias);
+            LOG(DEBUG) << "base_wlan_hal_whm::refresh_vap_info: vap_type = "
+                       << eVapType_str(vap_element.vap_type) << " for SSID=" << vap_element.ssid;
+
             m_ambiorix_cl.resolve_path(wifi_ssid_path, vap_extInfo.ssid_path);
             vap_extInfo.status = wbapi_utils::get_ap_status(ap_obj);
             LOG(INFO) << "status for " << ifname << " " << vap_extInfo.status;
diff --git a/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_apmanager.h b/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_apmanager.h
index e372ce122..4f1f9aae3 100644
--- a/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_apmanager.h
+++ b/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_apmanager.h
@@ -104,6 +104,7 @@ class cACTION_APMANAGER_JOINED_NOTIFICATION : public BaseClass
         bool add_channel_list(std::shared_ptr<cChannelList> ptr);
         std::shared_ptr<cChannelList> channel_list() { return m_channel_list_ptr; }
         sVapsList& vap_list();
+        sVapTypesList& vap_type_list();
         uint8_t& radio_max_bss();
         uint8_t& radio_rsn_override_support();
         void class_swap() override;
@@ -121,6 +122,7 @@ class cACTION_APMANAGER_JOINED_NOTIFICATION : public BaseClass
         bool m_lock_allocation__ = false;
         int m_lock_order_counter__ = 0;
         sVapsList* m_vap_list = nullptr;
+        sVapTypesList* m_vap_type_list = nullptr;
         uint8_t* m_radio_max_bss = nullptr;
         uint8_t* m_radio_rsn_override_support = nullptr;
 };
@@ -306,6 +308,7 @@ class cACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION : public BaseClass
             return (eActionOp_APMANAGER)(ACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION);
         }
         sVapsList& params();
+        sVapTypesList& vap_type_list();
         void class_swap() override;
         bool finalize() override;
         static size_t get_initial_size();
@@ -314,6 +317,7 @@ class cACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION : public BaseClass
         bool init();
         eActionOp_APMANAGER* m_action_op = nullptr;
         sVapsList* m_params = nullptr;
+        sVapTypesList* m_vap_type_list = nullptr;
 };
 
 class cACTION_APMANAGER_HOSTAP_CHANNEL_SWITCH_ACS_START : public BaseClass
diff --git a/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_common.h b/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_common.h
index 543b4ad81..a26ee532e 100644
--- a/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_common.h
+++ b/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_common.h
@@ -19,6 +19,7 @@
 #include <ostream>
 #include "beerocks/tlvf/beerocks_message_action.h"
 #include "bcl/beerocks_message_structs.h"
+#include "tlvf/common/eVapType.h"
 
 namespace beerocks_message {
 
@@ -63,6 +64,16 @@ typedef struct sVapInfo {
     }
 } __attribute__((packed)) sVapInfo;
 
+typedef struct sVapType {
+    int8_t vap_id;
+    eVapType vap_type;
+    void struct_swap(){
+        tlvf_swap(8*sizeof(eVapType), reinterpret_cast<uint8_t*>(&vap_type));
+    }
+    void struct_init(){
+    }
+} __attribute__((packed)) sVapType;
+
 typedef struct sSonConfig {
     uint8_t monitor_total_ch_load_notification_lo_th_percent;
     uint8_t monitor_total_ch_load_notification_hi_th_percent;
@@ -439,6 +450,20 @@ typedef struct sVapsList {
     }
 } __attribute__((packed)) sVapsList;
 
+typedef struct sVapTypesList {
+    sVapType vap_types[beerocks::IFACE_TOTAL_VAPS];
+    void struct_swap(){
+        for (size_t i = 0; i < beerocks::IFACE_TOTAL_VAPS; i++){
+            (vap_types[i]).struct_swap();
+        }
+    }
+    void struct_init(){
+            for (size_t i = 0; i < beerocks::IFACE_TOTAL_VAPS; i++) {
+                (vap_types[i]).struct_init();
+            }
+    }
+} __attribute__((packed)) sVapTypesList;
+
 typedef struct sArpMonitorData {
     sMacAddr mac;
     beerocks::net::sIpv4Addr ipv4;
diff --git a/common/beerocks/tlvf/AutoGenerated/src/beerocks/tlvf/beerocks_message_apmanager.cpp b/common/beerocks/tlvf/AutoGenerated/src/beerocks/tlvf/beerocks_message_apmanager.cpp
index 9669fa25c..d34efa0fa 100644
--- a/common/beerocks/tlvf/AutoGenerated/src/beerocks/tlvf/beerocks_message_apmanager.cpp
+++ b/common/beerocks/tlvf/AutoGenerated/src/beerocks/tlvf/beerocks_message_apmanager.cpp
@@ -287,6 +287,7 @@ std::shared_ptr<cChannelList> cACTION_APMANAGER_JOINED_NOTIFICATION::create_chan
         std::copy_n(src, move_length, dst);
     }
     m_vap_list = (sVapsList *)((uint8_t *)(m_vap_list) + len);
+    m_vap_type_list = (sVapTypesList *)((uint8_t *)(m_vap_type_list) + len);
     m_radio_max_bss = (uint8_t *)((uint8_t *)(m_radio_max_bss) + len);
     m_radio_rsn_override_support = (uint8_t *)((uint8_t *)(m_radio_rsn_override_support) + len);
     return std::make_shared<cChannelList>(src, getBuffRemainingBytes(src), m_parse__);
@@ -313,6 +314,7 @@ bool cACTION_APMANAGER_JOINED_NOTIFICATION::add_channel_list(std::shared_ptr<cCh
     m_channel_list_init = true;
     size_t len = ptr->getLen();
     m_vap_list = (sVapsList *)((uint8_t *)(m_vap_list) + len - ptr->get_initial_size());
+    m_vap_type_list = (sVapTypesList *)((uint8_t *)(m_vap_type_list) + len - ptr->get_initial_size());
     m_radio_max_bss = (uint8_t *)((uint8_t *)(m_radio_max_bss) + len - ptr->get_initial_size());
     m_radio_rsn_override_support = (uint8_t *)((uint8_t *)(m_radio_rsn_override_support) + len - ptr->get_initial_size());
     m_channel_list_ptr = ptr;
@@ -328,6 +330,10 @@ sVapsList& cACTION_APMANAGER_JOINED_NOTIFICATION::vap_list() {
     return (sVapsList&)(*m_vap_list);
 }
 
+sVapTypesList& cACTION_APMANAGER_JOINED_NOTIFICATION::vap_type_list() {
+    return (sVapTypesList&)(*m_vap_type_list);
+}
+
 uint8_t& cACTION_APMANAGER_JOINED_NOTIFICATION::radio_max_bss() {
     return (uint8_t&)(*m_radio_max_bss);
 }
@@ -343,6 +349,7 @@ void cACTION_APMANAGER_JOINED_NOTIFICATION::class_swap()
     m_cs_params->struct_swap();
     if (m_channel_list_ptr) { m_channel_list_ptr->class_swap(); }
     m_vap_list->struct_swap();
+    m_vap_type_list->struct_swap();
 }
 
 bool cACTION_APMANAGER_JOINED_NOTIFICATION::finalize()
@@ -378,6 +385,7 @@ size_t cACTION_APMANAGER_JOINED_NOTIFICATION::get_initial_size()
     class_size += sizeof(sNodeHostap); // params
     class_size += sizeof(sApChannelSwitch); // cs_params
     class_size += sizeof(sVapsList); // vap_list
+    class_size += sizeof(sVapTypesList); // vap_type_list
     class_size += sizeof(uint8_t); // radio_max_bss
     class_size += sizeof(uint8_t); // radio_rsn_override_support
     return class_size;
@@ -421,6 +429,12 @@ bool cACTION_APMANAGER_JOINED_NOTIFICATION::init()
         return false;
     }
     if (!m_parse__) { m_vap_list->struct_init(); }
+    m_vap_type_list = reinterpret_cast<sVapTypesList*>(m_buff_ptr__);
+    if (!buffPtrIncrementSafe(sizeof(sVapTypesList))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(sVapTypesList) << ") Failed!";
+        return false;
+    }
+    if (!m_parse__) { m_vap_type_list->struct_init(); }
     m_radio_max_bss = reinterpret_cast<uint8_t*>(m_buff_ptr__);
     if (!buffPtrIncrementSafe(sizeof(uint8_t))) {
         LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(uint8_t) << ") Failed!";
@@ -1008,10 +1022,15 @@ sVapsList& cACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION::params() {
     return (sVapsList&)(*m_params);
 }
 
+sVapTypesList& cACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION::vap_type_list() {
+    return (sVapTypesList&)(*m_vap_type_list);
+}
+
 void cACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION::class_swap()
 {
     tlvf_swap(8*sizeof(eActionOp_APMANAGER), reinterpret_cast<uint8_t*>(m_action_op));
     m_params->struct_swap();
+    m_vap_type_list->struct_swap();
 }
 
 bool cACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION::finalize()
@@ -1045,6 +1064,7 @@ size_t cACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION::get_initial_size(
 {
     size_t class_size = 0;
     class_size += sizeof(sVapsList); // params
+    class_size += sizeof(sVapTypesList); // vap_type_list
     return class_size;
 }
 
@@ -1060,6 +1080,12 @@ bool cACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION::init()
         return false;
     }
     if (!m_parse__) { m_params->struct_init(); }
+    m_vap_type_list = reinterpret_cast<sVapTypesList*>(m_buff_ptr__);
+    if (!buffPtrIncrementSafe(sizeof(sVapTypesList))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(sVapTypesList) << ") Failed!";
+        return false;
+    }
+    if (!m_parse__) { m_vap_type_list->struct_init(); }
     if (m_parse__) { class_swap(); }
     return true;
 }
diff --git a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml
index d3dae51bf..8ebc969cb 100755
--- a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml
+++ b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_apmanager.yaml
@@ -43,6 +43,7 @@ cACTION_APMANAGER_JOINED_NOTIFICATION:
   cs_params: sApChannelSwitch
   channel_list: cChannelList
   vap_list: sVapsList
+  vap_type_list: sVapTypesList
   radio_max_bss: uint8_t
   radio_rsn_override_support : uint8_t #bool
 
@@ -82,6 +83,7 @@ cACTION_APMANAGER_HOSTAP_GENERATE_CLIENT_ASSOCIATION_NOTIFICATIONS_REQUEST:
 cACTION_APMANAGER_HOSTAP_VAPS_LIST_UPDATE_NOTIFICATION:
   _type: class
   params: sVapsList
+  vap_type_list: sVapTypesList
 
 cACTION_APMANAGER_HOSTAP_CHANNEL_SWITCH_ACS_START:
   _type: class
diff --git a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml
index ea8e42330..8d19883b9 100755
--- a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml
+++ b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml
@@ -1,7 +1,7 @@
 #
 ---
 _include:
-  { beerocks/tlvf/beerocks_message_action.h, bcl/beerocks_message_structs.h }
+  { beerocks/tlvf/beerocks_message_action.h, bcl/beerocks_message_structs.h, tlvf/common/eVapType.h }
 
 _namespace: beerocks_message
 
@@ -51,6 +51,11 @@ sVapInfo:
     _type: int8_t
   ap_mld_mac: sMacAddr
 
+sVapType:
+  _type: struct
+  vap_id: int8_t
+  vap_type: eVapType
+
 sSonConfig:
   _type: struct
   monitor_total_ch_load_notification_lo_th_percent: uint8_t
@@ -333,6 +338,12 @@ sVapsList:
     _type: sVapInfo
     _length: ["beerocks::IFACE_TOTAL_VAPS"]
 
+sVapTypesList:
+  _type: struct
+  vap_types:
+    _type: sVapType
+    _length: ["beerocks::IFACE_TOTAL_VAPS"]
+
 sArpMonitorData:
   _type: struct
   mac: sMacAddr
diff --git a/framework/platform/bpl/cfg/dm/bpl_cfg_wifi.cpp b/framework/platform/bpl/cfg/dm/bpl_cfg_wifi.cpp
index b31daab3d..1402ad91d 100644
--- a/framework/platform/bpl/cfg/dm/bpl_cfg_wifi.cpp
+++ b/framework/platform/bpl/cfg/dm/bpl_cfg_wifi.cpp
@@ -18,6 +18,8 @@
 #include <tlvf/WSC/eWscAuth.h>
 #include <tlvf/WSC/eWscEncr.h>
 
+#include <tlvf/common/eVapType.h>
+
 #include "wbapi_utils.h"
 
 #include "bpl_cfg_pwhm.h"
@@ -196,6 +198,12 @@ bool bpl_cfg_get_wireless_settings(std::list<son::wireless_utils::sBssInfoConf>
 
         configuration.mld_id = std::to_string(mld_id);
 
+        // Reading CustomAlias and deducing VAP type
+        const std::string custom_alias = wbapi_utils::get_custom_alias(ap);
+        configuration.vap_type         = wbapi_utils::vap_type_from_custom_alias(custom_alias);
+        LOG(DEBUG) << "bpl_cfg_get_wireless_settings: vap_type is "
+                   << eVapType_str(configuration.vap_type) << " for SSID=" << configuration.ssid;
+
         bool ap_enable = false;
         ap.read_child(ap_enable, "Enable");
         if (ap_enable && bpl_cfg_get_wifi_credentials(iface, configuration)) {
diff --git a/framework/platform/wbapi/include/wbapi_utils.h b/framework/platform/wbapi/include/wbapi_utils.h
index 1342417d2..1a972083f 100644
--- a/framework/platform/wbapi/include/wbapi_utils.h
+++ b/framework/platform/wbapi/include/wbapi_utils.h
@@ -13,8 +13,10 @@
 #include <bcl/beerocks_defines.h>
 #include <bcl/son/son_wireless_utils.h>
 #include <easylogging++.h>
+
 #include <tlvf/WSC/eWscAuth.h>
 #include <tlvf/WSC/eWscEncr.h>
+#include <tlvf/common/eVapType.h>
 
 namespace beerocks {
 namespace wbapi {
@@ -91,6 +93,19 @@ public:
      */
     static WSC::eWscEncr encryption_type_from_auth(WSC::eWscAuth auth);
 
+    /**
+     * @brief Convert custom_alias -> vap_type.
+     *
+     * Examples:
+     *   "vap5ghome", "homevap24g", "vaphome6g" -> HOME
+     *   "vap5gguest", "guestvap24g", "6gguestvap" -> GUEST
+     *   "vap5gbackhaul", "backhaulvap24g", "6gvapbackhaul" -> BACKHAUL
+     *
+     * @param custom_alias  Arbitrary alias string (e.g. "vap5ghome").
+     * @return eVapType  Detected type or OTHER if no match.
+     */
+    static eVapType vap_type_from_custom_alias(const std::string &custom_alias);
+
     /**
      * @brief get amxc var object id from the object path.
      * eg: "RootObj.TemplateObj.id" or "RootObj.TemplateObj.id."
@@ -174,6 +189,11 @@ public:
     */
     static std::string search_path_ssid_by_alias(const std::string &alias);
 
+    /**
+     * @brief return search path of SSID object instance by CustomAlias
+    */
+    static std::string search_path_ssid_by_custom_alias(const std::string &custom_alias);
+
     /**
      * @brief return search path of SSID object instance by SSID.
      *
@@ -312,6 +332,11 @@ public:
      */
     static std::string get_ap_iface(const AmbiorixVariant &obj);
 
+    /**
+     * @brief get CustomAlias of AccessPoint object data
+     */
+    static std::string get_custom_alias(const AmbiorixVariant &obj);
+
     /**
      * @brief get status of AccessPoint object data
      */
diff --git a/framework/platform/wbapi/wbapi_utils.cpp b/framework/platform/wbapi/wbapi_utils.cpp
index 47bab29ad..551c1655b 100644
--- a/framework/platform/wbapi/wbapi_utils.cpp
+++ b/framework/platform/wbapi/wbapi_utils.cpp
@@ -201,6 +201,26 @@ WSC::eWscEncr wbapi_utils::encryption_type_from_auth(WSC::eWscAuth auth)
     return WSC::eWscEncr::WSC_ENCR_AES;
 }
 
+eVapType wbapi_utils::vap_type_from_custom_alias(const std::string &custom_alias)
+{
+    if (custom_alias.find("home") != std::string::npos)
+        return eVapType::HOME;
+    else if (custom_alias.find("guest") != std::string::npos)
+        return eVapType::GUEST;
+    else if (custom_alias.find("video") != std::string::npos)
+        return eVapType::VIDEO;
+    else if (custom_alias.find("backhaul") != std::string::npos)
+        return eVapType::BACKHAUL;
+    else if (custom_alias.find("hotspot") != std::string::npos)
+        return eVapType::HOTSPOT;
+    else if (custom_alias.find("staff") != std::string::npos)
+        return eVapType::STAFF;
+    else if (custom_alias.find("isolated") != std::string::npos)
+        return eVapType::ISOLATED;
+    else
+        return eVapType::OTHER;
+}
+
 int wbapi_utils::get_object_id(const std::string &object_path)
 {
     auto str = object_path;
@@ -270,6 +290,11 @@ std::string wbapi_utils::search_path_ssid_by_alias(const std::string &alias)
     return search_path_ssid() + "[Alias == '" + alias + "'].";
 }
 
+std::string wbapi_utils::search_path_ssid_by_custom_alias(const std::string &custom_alias)
+{
+    return search_path_ssid() + "[CustomAlias == '" + custom_alias + "'].";
+}
+
 std::string wbapi_utils::search_path_ssid_by_ssid_and_radio(const std::string &ssid,
                                                             const std::string &radio_path)
 {
@@ -435,6 +460,21 @@ std::string wbapi_utils::get_ap_iface(const AmbiorixVariant &obj)
     return value;
 }
 
+std::string wbapi_utils::get_custom_alias(const AmbiorixVariant &obj)
+{
+    std::string value;
+    if (!obj.read_child(value, "CustomAlias")) {
+        LOG(ERROR) << "wbapi_utils::get_custom_alias: Failed to read CustomAlias";
+        return {};
+    }
+
+    if (value.empty()) {
+        LOG(WARNING) << "wbapi_utils::get_custom_alias: CustomAlias is empty";
+    }
+
+    return value;
+}
+
 std::string wbapi_utils::get_ap_status(const AmbiorixVariant &obj)
 {
     std::string value;
diff --git a/framework/tlvf/AutoGenerated/include/tlvf/WSC/WSC_Attributes.h b/framework/tlvf/AutoGenerated/include/tlvf/WSC/WSC_Attributes.h
index bb052972d..0f7dffa16 100644
--- a/framework/tlvf/AutoGenerated/include/tlvf/WSC/WSC_Attributes.h
+++ b/framework/tlvf/AutoGenerated/include/tlvf/WSC/WSC_Attributes.h
@@ -21,13 +21,14 @@
 #include <tlvf/BaseClass.h>
 #include <tlvf/ClassList.h>
 #include <tuple>
+#include "tlvf/common/sMacAddr.h"
+#include "tlvf/common/eVapType.h"
 #include "tlvf/WSC/eWscLengths.h"
 #include "tlvf/WSC/eWscAuth.h"
 #include "tlvf/WSC/eWscVendorId.h"
 #include "tlvf/WSC/eWscVendorExt.h"
 #include "tlvf/WSC/eWscDev.h"
 #include "tlvf/WSC/eWscWfaVendorExtSubelement.h"
-#include "tlvf/common/sMacAddr.h"
 #include "tlvf/WSC/eWscAttributes.h"
 #include "tlvf/WSC/eWscEncr.h"
 #include "tlvf/WSC/eWscValues8.h"
@@ -232,6 +233,7 @@ class cEncryptedSettingsPayload : public BaseClass
         uint8_t& hidden_ssid();
         int8_t& bss_index();
         uint8_t& additional_auth();
+        eVapType& vap_type();
         void class_swap() override;
         bool finalize() override;
         static size_t get_initial_size();
@@ -255,6 +257,7 @@ class cEncryptedSettingsPayload : public BaseClass
         uint8_t* m_hidden_ssid = nullptr;
         int8_t* m_bss_index = nullptr;
         uint8_t* m_additional_auth = nullptr;
+        eVapType* m_vap_type = nullptr;
 };
 
 class cWscAttrEncryptedSettings : public BaseClass
diff --git a/framework/tlvf/AutoGenerated/include/tlvf/common/eVapType.h b/framework/tlvf/AutoGenerated/include/tlvf/common/eVapType.h
new file mode 100644
index 000000000..5f81cc2bc
--- /dev/null
+++ b/framework/tlvf/AutoGenerated/include/tlvf/common/eVapType.h
@@ -0,0 +1,73 @@
+///////////////////////////////////////
+// AUTO GENERATED FILE - DO NOT EDIT //
+///////////////////////////////////////
+
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _TLVF_COMMON_EVAPTYPE_H_
+#define _TLVF_COMMON_EVAPTYPE_H_
+
+#include <cstddef>
+#include <stdint.h>
+#include <tlvf/swap.h>
+#include <ostream>
+enum class eVapType : uint8_t {
+    HOME = 0x0,
+    GUEST = 0x1,
+    VIDEO = 0x2,
+    BACKHAUL = 0x3,
+    HOTSPOT = 0x4,
+    STAFF = 0x5,
+    ISOLATED = 0x6,
+    OTHER = 0xff,
+};
+// Enum AutoPrint generated code snippet begining- DON'T EDIT!
+// clang-format off
+static const char *eVapType_str(eVapType enum_value) {
+    switch (enum_value) {
+    case eVapType::HOME:     return "eVapType::HOME";
+    case eVapType::GUEST:    return "eVapType::GUEST";
+    case eVapType::VIDEO:    return "eVapType::VIDEO";
+    case eVapType::BACKHAUL: return "eVapType::BACKHAUL";
+    case eVapType::HOTSPOT:  return "eVapType::HOTSPOT";
+    case eVapType::STAFF:    return "eVapType::STAFF";
+    case eVapType::ISOLATED: return "eVapType::ISOLATED";
+    case eVapType::OTHER:    return "eVapType::OTHER";
+    }
+    static std::string out_str = std::to_string(int(enum_value));
+    return out_str.c_str();
+}
+inline std::ostream &operator<<(std::ostream &out, eVapType value) { return out << eVapType_str(value); }
+// clang-format on
+// Enum AutoPrint generated code snippet end
+class eVapTypeValidate {
+public:
+    static bool check(uint8_t value) {
+        bool ret = false;
+        switch (value) {
+        case 0x0:
+        case 0x1:
+        case 0x2:
+        case 0x3:
+        case 0x4:
+        case 0x5:
+        case 0x6:
+        case 0xff:
+                ret = true;
+                break;
+            default:
+                ret = false;
+                break;
+        }
+        return ret;
+    }
+};
+
+
+#endif //_TLVF/COMMON_EVAPTYPE_H_
diff --git a/framework/tlvf/AutoGenerated/src/tlvf/WSC/WSC_Attributes.cpp b/framework/tlvf/AutoGenerated/src/tlvf/WSC/WSC_Attributes.cpp
index b414237a7..7f8b2d548 100644
--- a/framework/tlvf/AutoGenerated/src/tlvf/WSC/WSC_Attributes.cpp
+++ b/framework/tlvf/AutoGenerated/src/tlvf/WSC/WSC_Attributes.cpp
@@ -391,6 +391,7 @@ bool cEncryptedSettingsPayload::alloc_ssid(size_t count) {
     m_hidden_ssid = (uint8_t *)((uint8_t *)(m_hidden_ssid) + len);
     m_bss_index = (int8_t *)((uint8_t *)(m_bss_index) + len);
     m_additional_auth = (uint8_t *)((uint8_t *)(m_additional_auth) + len);
+    m_vap_type = (eVapType *)((uint8_t *)(m_vap_type) + len);
     m_ssid_idx__ += count;
     *m_ssid_length += count;
     if (!buffPtrIncrementSafe(len)) {
@@ -480,6 +481,7 @@ bool cEncryptedSettingsPayload::alloc_network_key(size_t count) {
     m_hidden_ssid = (uint8_t *)((uint8_t *)(m_hidden_ssid) + len);
     m_bss_index = (int8_t *)((uint8_t *)(m_bss_index) + len);
     m_additional_auth = (uint8_t *)((uint8_t *)(m_additional_auth) + len);
+    m_vap_type = (eVapType *)((uint8_t *)(m_vap_type) + len);
     m_network_key_idx__ += count;
     *m_network_key_length += count;
     if (!buffPtrIncrementSafe(len)) {
@@ -513,6 +515,10 @@ uint8_t& cEncryptedSettingsPayload::additional_auth() {
     return (uint8_t&)(*m_additional_auth);
 }
 
+eVapType& cEncryptedSettingsPayload::vap_type() {
+    return (eVapType&)(*m_vap_type);
+}
+
 void cEncryptedSettingsPayload::class_swap()
 {
     tlvf_swap(16, reinterpret_cast<uint8_t*>(m_ssid_type));
@@ -566,6 +572,7 @@ size_t cEncryptedSettingsPayload::get_initial_size()
     class_size += sizeof(uint8_t); // hidden_ssid
     class_size += sizeof(int8_t); // bss_index
     class_size += sizeof(uint8_t); // additional_auth
+    class_size += sizeof(eVapType); // vap_type
     return class_size;
 }
 
@@ -663,6 +670,12 @@ bool cEncryptedSettingsPayload::init()
         LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(uint8_t) << ") Failed!";
         return false;
     }
+    m_vap_type = reinterpret_cast<eVapType*>(m_buff_ptr__);
+    if (!m_parse__) *m_vap_type = eVapType::OTHER;
+    if (!buffPtrIncrementSafe(sizeof(eVapType))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(eVapType) << ") Failed!";
+        return false;
+    }
     if (m_parse__) { class_swap(); }
     return true;
 }
diff --git a/framework/tlvf/src/include/tlvf/WSC/m2.h b/framework/tlvf/src/include/tlvf/WSC/m2.h
index 8f8e06cf9..4720ad57c 100644
--- a/framework/tlvf/src/include/tlvf/WSC/m2.h
+++ b/framework/tlvf/src/include/tlvf/WSC/m2.h
@@ -15,8 +15,9 @@ namespace WSC {
 namespace vendor_extension {
 namespace airties {
 
-constexpr int VENDOR_HIDE_SSID = 0x80;
-constexpr int VENDOR_BSS_CFG   = 0x02;
+constexpr uint8_t VENDOR_HIDE_SSID = 0x80;
+constexpr uint8_t VENDOR_BSS_CFG   = 0x02;
+constexpr uint8_t VENDOR_VAP_TYPE  = 0x00;
 
 } // namespace airties
 } // namespace vendor_extension
@@ -73,6 +74,9 @@ public:
 
         /* Hidden SSID flag (Airties Vendor Extension subelement: hidden_ssid) */
         bool hidden_ssid = false;
+
+        /* WSC VAP Type */
+        eVapType vap_type = eVapType::OTHER;
     };
 
     m2(uint8_t *buff, size_t buff_len, bool parse) : WscAttrList(buff, buff_len, parse) {}
@@ -142,6 +146,35 @@ public:
         }
         return false;
     };
+    eVapType vap_type() const
+    {
+        for (auto &vendor_ext_attr : getAttrList<WSC::cWscAttrVendorExtension>()) {
+            if ((WSC::eWscVendorId::WSC_VENDOR_ID_AIRTIES_1 != vendor_ext_attr->vendor_id_0()) ||
+                (WSC::eWscVendorId::WSC_VENDOR_ID_AIRTIES_2 != vendor_ext_attr->vendor_id_1()) ||
+                (WSC::eWscVendorId::WSC_VENDOR_ID_AIRTIES_3 != vendor_ext_attr->vendor_id_2())) {
+                continue;
+            }
+
+            const auto *data = vendor_ext_attr->vendor_data();
+            const size_t len = vendor_ext_attr->vendor_data_length();
+
+            // Need at least 2 bytes: [subelem_id][value]
+            if (len >= 2 && data[0] == WSC::vendor_extension::airties::VENDOR_VAP_TYPE) {
+
+                const uint8_t raw = data[1];
+
+                if (!eVapTypeValidate::check(raw)) {
+                    LOG(WARNING) << "Invalid VAP type received in vendor extension: " << int(raw)
+                                 << " -> using OTHER";
+                    return eVapType::OTHER;
+                }
+
+                return static_cast<eVapType>(raw);
+            }
+        }
+
+        return eVapType::OTHER;
+    }
 };
 
 } // namespace WSC
diff --git a/framework/tlvf/src/src/WSC/m2.cpp b/framework/tlvf/src/src/WSC/m2.cpp
index 0d6605855..c543cae8c 100644
--- a/framework/tlvf/src/src/WSC/m2.cpp
+++ b/framework/tlvf/src/src/WSC/m2.cpp
@@ -193,34 +193,60 @@ bool m2::init(const config &cfg, bool bss_index_support)
         return false;
     }
 
-    auto vendor_ext_attr = addAttr<cWscAttrVendorExtension>();
-    if (!vendor_ext_attr) {
-        TLVF_LOG(ERROR) << "addAttr<cWscAttrVendorExtension> failed";
-        return false;
-    }
+    // WFA Vendor Extension
+    {
+        auto vendor_ext_attr = addAttr<cWscAttrVendorExtension>();
+        if (!vendor_ext_attr) {
+            TLVF_LOG(ERROR) << "addAttr<cWscAttrVendorExtension> failed";
+            return false;
+        }
 
-    // WFA Vendor Data
-    const size_t vendor_data_size = sizeof(sWscWfaVendorExtSubelementVersion2);
-    if (!vendor_ext_attr->alloc_vendor_data(vendor_data_size)) {
-        LOG(ERROR) << "Failed to allocate vendor data [" << vendor_data_size << "]!";
-        return false;
-    }
-    auto vendor_data = vendor_ext_attr->vendor_data();
+        // WFA Vendor Data
+        const size_t vendor_data_size = sizeof(sWscWfaVendorExtSubelementVersion2);
+        if (!vendor_ext_attr->alloc_vendor_data(vendor_data_size)) {
+            LOG(ERROR) << "Failed to allocate vendor data [" << vendor_data_size << "]!";
+            return false;
+        }
+        auto vendor_data = vendor_ext_attr->vendor_data();
 
-    // WFA Vendor Extension Subelement at #0: Version2
-    sWscWfaVendorExtSubelementVersion2 version2{eWscWfaVendorExtSubelement::VERSION2, 0x01,
-                                                eWscVendorExtVersionIE::WSC_VERSION2};
-    std::copy_n(reinterpret_cast<uint8_t *>(&version2), sizeof(version2), vendor_data);
+        // WFA Vendor Extension Subelement at #0: Version2
+        sWscWfaVendorExtSubelementVersion2 version2{eWscWfaVendorExtSubelement::VERSION2, 0x01,
+                                                    eWscVendorExtVersionIE::WSC_VERSION2};
+        std::copy_n(reinterpret_cast<uint8_t *>(&version2), sizeof(version2), vendor_data);
 
-    auto encrypted_settings = addAttr<cWscAttrEncryptedSettings>();
-    if (!encrypted_settings) {
-        TLVF_LOG(ERROR) << "addAttr<cWscAttrEncryptedSettings> failed";
-        return false;
+        auto encrypted_settings = addAttr<cWscAttrEncryptedSettings>();
+        if (!encrypted_settings) {
+            TLVF_LOG(ERROR) << "addAttr<cWscAttrEncryptedSettings> failed";
+            return false;
+        }
+        std::copy_n(cfg.iv, WSC_ENCRYPTED_SETTINGS_IV_LENGTH, encrypted_settings->iv());
+        encrypted_settings->alloc_encrypted_settings(cfg.encrypted_settings.size());
+        std::copy_n(cfg.encrypted_settings.data(), encrypted_settings->encrypted_settings_length(),
+                    encrypted_settings->encrypted_settings());
+    }
+
+    // Airties Vendor Extension
+    {
+        auto vendor_ext_attr = addAttr<cWscAttrVendorExtension>();
+        if (!vendor_ext_attr) {
+            TLVF_LOG(ERROR) << "addAttr<cWscAttrVendorExtension> (Airties) failed";
+            return false;
+        }
+        vendor_ext_attr->vendor_id_0() = WSC::eWscVendorId::WSC_VENDOR_ID_AIRTIES_1;
+        vendor_ext_attr->vendor_id_1() = WSC::eWscVendorId::WSC_VENDOR_ID_AIRTIES_2;
+        vendor_ext_attr->vendor_id_2() = WSC::eWscVendorId::WSC_VENDOR_ID_AIRTIES_3;
+
+        // Payload: [ tag, value ]
+        const uint8_t payload[2] = {
+            static_cast<uint8_t>(WSC::vendor_extension::airties::VENDOR_VAP_TYPE),
+            static_cast<uint8_t>(cfg.vap_type),
+        };
+
+        if (!vendor_ext_attr->set_vendor_data(payload, sizeof(payload))) {
+            TLVF_LOG(ERROR) << "set_vendor_data (Airties) failed";
+            return false;
+        }
     }
-    std::copy_n(cfg.iv, WSC_ENCRYPTED_SETTINGS_IV_LENGTH, encrypted_settings->iv());
-    encrypted_settings->alloc_encrypted_settings(cfg.encrypted_settings.size());
-    std::copy_n(cfg.encrypted_settings.data(), encrypted_settings->encrypted_settings_length(),
-                encrypted_settings->encrypted_settings());
 
     // For retro compatibility, don't send bss_index for agent not supporting RSN overriding
     if (bss_index_support) {
diff --git a/framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml b/framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml
index d8b8a87f8..4cb61e6c7 100644
--- a/framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml
+++ b/framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml
@@ -1,13 +1,14 @@
 #
 ---
 _include: {
+  tlvf/common/sMacAddr.h,
+  tlvf/common/eVapType.h,
   tlvf/WSC/eWscLengths.h,
   tlvf/WSC/eWscAuth.h,
   tlvf/WSC/eWscVendorId.h,
   tlvf/WSC/eWscVendorExt.h,
   tlvf/WSC/eWscDev.h,
   tlvf/WSC/eWscWfaVendorExtSubelement.h,
-  tlvf/common/sMacAddr.h,
   tlvf/WSC/eWscAttributes.h,
   tlvf/WSC/eWscEncr.h,
   tlvf/WSC/eWscValues8.h,
@@ -156,6 +157,9 @@ cEncryptedSettingsPayload:
   additional_auth:
     _type: uint8_t
     _value: 0
+  vap_type:
+    _type: eVapType
+    _value: eVapType::OTHER
 
 cWscAttrEncryptedSettings:
   _type: class
diff --git a/framework/tlvf/yaml/tlvf/common/eVapType.yaml b/framework/tlvf/yaml/tlvf/common/eVapType.yaml
new file mode 100644
index 000000000..162871be5
--- /dev/null
+++ b/framework/tlvf/yaml/tlvf/common/eVapType.yaml
@@ -0,0 +1,15 @@
+#
+---
+
+eVapType:
+  _type: enum_class
+  _enum_storage: uint8_t
+  HOME: 0x00
+  GUEST: 0x01
+  VIDEO: 0x02
+  BACKHAUL: 0x03
+  HOTSPOT: 0x04
+  STAFF: 0x05
+  ISOLATED: 0x06
+  # 7-254 = UNSPECIFIED
+  OTHER: 0xFF
-- 
2.34.1

